zerver/lib/subscription_info.py

```diff
@@ -336,23 +337,50 @@ def validate_user_access_to_subscribers(user_profile: UserProfile | None, stream
     * The stream is invite only, requesting_user is passed, and that user
       does not subscribe to the stream.
     """
+    user_recursive_group_ids: set[int] = set()
+    # Optimization for the organization administrator code path. We
+    # don't explicitly grant realm admins this permission, but admins
+    # implicitly have the can_administer_channel_group permission for
+    # all channels. user_recursive_group_ids is used to check the
+    # membership of the current user in can_administer_channel_group
+    # which we don't need to calculate in case of a realm admin.
+    if user_profile and not user_profile.is_realm_admin:
+        user_recursive_group_ids = set(
+            get_recursive_membership_groups(user_profile).values_list("id", flat=True)
+        )
```

Comment:
Let's open a follow-up issue to track optimizing `validate_user_access_to_subscribers`; I think it's fine to skip for this PR, and we will want it for the release, but it should be an easy follow-up.

Reply:
#33374